service/controller/credential/controller.go

```diff
@@ -0,0 +1,125 @@
+package credential
+
+import (
+	"context"
+
+	"github.com/giantswarm/certs/v3/pkg/certs"
+	"github.com/giantswarm/k8sclient/v5/pkg/k8sclient"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/giantswarm/operatorkit/v4/pkg/controller"
+	"github.com/giantswarm/operatorkit/v4/pkg/resource"
+	"github.com/giantswarm/operatorkit/v4/pkg/resource/wrapper/metricsresource"
+	"github.com/giantswarm/operatorkit/v4/pkg/resource/wrapper/retryresource"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/giantswarm/azure-operator/v5/pkg/label"
+	"github.com/giantswarm/azure-operator/v5/service/controller/credential/handler/azureclusteridentity"
+
+	"github.com/giantswarm/azure-operator/v5/pkg/project"
+	"github.com/giantswarm/azure-operator/v5/service/controller/controllercontext"
+)
+
+type ControllerConfig struct {
+	K8sClient k8sclient.Interface
+	Logger    micrologger.Logger
+	SentryDSN string
+}
+
+func NewController(config ControllerConfig) (*controller.Controller, error) {
+	var err error
+
+	var certsSearcher *certs.Searcher
+	{
+		c := certs.Config{
+			K8sClient: config.K8sClient.K8sClient(),
+			Logger:    config.Logger,
+		}
+
+		certsSearcher, err = certs.NewSearcher(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var resources []resource.Interface
+	{
+		resources, err = newCredentialResources(config, certsSearcher)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var operatorkitController *controller.Controller
+	{
+		c := controller.Config{
+			InitCtx: func(ctx context.Context, obj interface{}) (context.Context, error) {
+				return controllercontext.NewContext(ctx, controllercontext.Context{}), nil
+			},
+			K8sClient: config.K8sClient,
+			Logger:    config.Logger,
+			// Name is used to compute finalizer names. This results in something
+			// like operatorkit.giantswarm.io/azure-operator-credential-controller.
+			Name: project.Name() + "-credential-controller",
+			NewRuntimeObjectFunc: func() runtime.Object {
+				return new(v1.Secret)
+			},
+			Resources: resources,
+			Selector: labels.SelectorFromSet(map[string]string{
+				label.App: "credentiald",
+			}),
```

Comment:
I'd drop the separate controller here and just run the handler from `AzureClusterController` as part of the general reconciliation, before executing the upcoming handler that patches `AzureClusterIdentity` into it.

This implementation now causes the actual live race between operators because there's not version matching here (and there shouldn't be) so all operators jump in as soon as they can to update the related CRs. There is also potential for bigger problem in case we would ever implement any kind of garbage collection. The older operator possibly running would drop the finalizer immediately while the newer operator would expect it to be kept until dependent subresources are deleted. This would be very wrong.

The functionality is very simple in the end so it's fine to be part of `AzureClusterController` IMO.

Reply:
OK, agreed will move it

Reply:
done. What's missing now is the cleanup of `AzureCluterIdentity` objects that have been created from a legacy secret that was deleted. How do you suggest doing it?

Reply:
I'd do something roughly like this (context: handler reconciling `AzureCluster`):
- Compute `legacySecretName` from referenced `AzureClusterIdentity`
- Find if `legacySecretName` still exists.
- If not, list all `AzureClusters` and check if given `AzureClusterIdentity` is referenced from there.
- If not, delete the given `AzureClusterIdentity` and set `AzureCluster` CR `IdentityRef` to nil.

This way we should minimize the chance that some other cluster is about-to-use the given to-be-deleted `AzureClusterIdentity`.

Reply:
done